beater/java_attacher/java_attacher.go

```diff
@@ -25,50 +25,369 @@ import (
+	regex, err := regexp.Compile(regexS)
+	if err != nil {
+		j.logger.Errorf("invalid regex for the `%v` argument: %v", argumentName, err)
+		return
+	}
+	j.addDiscoveryRule(cmdLineDiscoveryRule{regex: regex, isIncludeRule: isIncludeRule, argumentName: argumentName})
+}
+
+func (j *JavaAttacher) findFirstMatch(jvm *JvmDetails) discoveryRule {
+	for _, rule := range j.discoveryRules {
+		if rule.match(jvm) {
+			return rule
+		}
+	}
+	return nil
+}
+
+func (j *JavaAttacher) Run(ctx context.Context) error {
+	if !j.enabled {
+		return fmt.Errorf("java attacher is disabled")
+	}
+
+	if runtime.GOOS == "windows" {
+		// in Windows, we run the attacher by a JVM available to the current user in continuous mode and return immediately
+		if err := j.discoverJavaExecutable(); err != nil {
+			return err
+		}
+		return j.attach(ctx, &JvmDetails{})
+	}
+
+	// non-Windows - run discovery and attachment until context is closed
+	for {
+		jvms, err := j.discoverJvmsForAttachment(ctx)
+		if err != nil {
+			j.logger.Infof("error during JVMs discovery: %v", err)
+		} else {
+			if err := j.executeForEachJvm(ctx, jvms, j.attach, true, 30*time.Second); err != nil {
+				j.logger.Error(err)
+			}
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			time.Sleep(time.Second)
+		}
+	}
+}
+
+func (j *JavaAttacher) discoverJavaExecutable() error {
+	if j.javaBin == "" {
 		if jh := os.Getenv("JAVA_HOME"); jh != "" {
-			cfg.JavaBin = filepath.Join(jh, "/bin/java")
+			j.javaBin = filepath.Join(jh, "/bin/java")
 		} else {
 			bin, err := exec.LookPath("java")
 			if err != nil {
-				return JavaAttacher{}, fmt.Errorf("no java binary found: %v", err)
+				return fmt.Errorf("no java binary found: %v", err)
 			}
-			cfg.JavaBin = bin
+			j.javaBin = bin
 		}
 	} else {
 		// Ensure we're using the correct separators for the system
 		// running apm-server
-		cfg.JavaBin = filepath.FromSlash(cfg.JavaBin)
+		j.javaBin = filepath.FromSlash(j.javaBin)
 	}
-	logger := logp.NewLogger("java-attacher")
-	if _, err := os.Stat(javaAttacher); err != nil {
-		return JavaAttacher{}, err
+	return nil
+}
+
+// this function blocks until discovery has ended, an error occurred or the context had been cancelled
+func (j *JavaAttacher) discoverJvmsForAttachment(ctx context.Context) (map[string]*JvmDetails, error) {
+	jvms, err := j.discoverAllRunningJavaProcesses(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// remove stale processes from the cache
+	for pid := range encounteredJvmCache {
+		_, found := jvms[pid]
+		if !found {
+			delete(encounteredJvmCache, pid)
+		}
+	}
+
+	// trying to improve start time accuracy - worth to consider optimizing as this runs every time the loop runs (1 second or so)
+	if err := j.executeForEachJvm(ctx, jvms, j.obtainAccurateStartTime, false, time.Second); err != nil {
+		j.logger.Infof("finding accurate start time for %v jvms did not finish successfully: %v", len(jvms), err)
```

Comment:
**issue**: because this is logging an error, it should use the error level (`Errorf`)

Note: it applies to other lines in this file.

Reply:
As above.